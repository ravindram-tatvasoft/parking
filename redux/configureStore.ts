import { combineReducers, createStore } from 'redux'
import { persistStore, persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage' // defaults to localStorage for web
import parkingSlotSlice from './reducer/parking-slot-slice'
import AsyncStorage from '@react-native-async-storage/async-storage';


const persistConfig = {
    key: 'root',
    storage: AsyncStorage,
}

const rootReducer = combineReducers({
    parkingSlot: parkingSlotSlice,
})

const persistedReducer = persistReducer(persistConfig, rootReducer)

let store = createStore(persistedReducer)
let persistor = persistStore(store)

export {
    store, persistor
}
