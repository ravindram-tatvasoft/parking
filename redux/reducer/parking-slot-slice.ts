import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { SlotListType } from '../../types/slot-list-tyle'

export interface CounterState {
  totalParking: number
  slotList: SlotListType[]
}

const initialState: CounterState = {
    totalParking: 0,
    slotList: []
}

export const parkingSlotSlice = createSlice({
  name: 'parkingSlot',
  initialState,
  reducers: {
    setTotalParking: (state, action) => {
      state.totalParking = action.payload
    },
    setParkingSlotList: (state, action) => {
      state.slotList = action.payload
    },
  },
})

// Action creators are generated for each case reducer function
export const { setTotalParking, setParkingSlotList  } = parkingSlotSlice.actions

export default parkingSlotSlice.reducer