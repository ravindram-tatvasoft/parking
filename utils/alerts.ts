import { Alert } from "react-native"

export const showAlert = (message: string): void => {
    Alert.alert(message)
}

