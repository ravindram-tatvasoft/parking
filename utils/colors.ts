export const colors = {
    primary : "#246EE9",
    greyLight: "#d5d5d5",
    green: "green",
    red: "red"
}