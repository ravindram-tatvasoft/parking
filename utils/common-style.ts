import { StyleSheet } from "react-native";

const commonStyles = StyleSheet.create({
    container: {
      flex: 1,
      width: "100%",
    },
    viewCenter: {
      flex: 1,
      alignItems:'center',
      justifyContent: "center"
    },
    paddingHoriontal: {
      paddingHorizontal: 20
    }
  });
  
  export {
    commonStyles
  }