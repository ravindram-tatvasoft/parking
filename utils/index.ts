export * from './strings'
export * from './colors'
export * from './common-style'
export * from './utils'