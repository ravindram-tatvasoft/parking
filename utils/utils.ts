import moment from 'moment'

const msToTime = (milli: number) => {
  return moment(milli).format("HH:mm:ss")
}

const getDifferenceDateString = (startTime: string, endTime: string): string => {
  var startDate = moment(startTime);
  var endDate = moment(endTime);

  var totalSec = endDate.diff(startDate, 'seconds');
  const time = moment().startOf('day')
        .seconds(totalSec)
        .format('HH@mm').split("@");
    if(parseInt(time[0]) > 0){
      return `${time[0]} hours ${time[1]} minutes`
    } else{
      return `${time[1]} minutes`
    }
}

const validateCarNo = (carNo: string): boolean => {
  var pattern = /[A-Z][A-Z] [0-9][0-9] [A-Z][A-Z] [0-9][0-9][0-9][0-9]$/i;
  var re = new RegExp(pattern);
  return re.test(carNo)
}

const getDifferenceDate = (startTime: string, endTime: string): any => {
  var startDate = moment(startTime);
  var endDate = moment(endTime);

  const h = endDate.diff(startDate, 'hours')
  const m = endDate.diff(startDate, 'minutes')
  return { hour: h, minute: m }
}

export {
  msToTime,
  getDifferenceDate,
  getDifferenceDateString,
  validateCarNo
}