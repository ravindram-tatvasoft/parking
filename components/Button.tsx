import {StyleSheet, TouchableOpacity, Text, TouchableOpacityProps, ViewStyle } from "react-native"
import { colors } from "../utils"

export interface ButtonProps extends TouchableOpacityProps {
    name: string,
    onPressButton: () => void,
    testId: string
    rootStyle?: ViewStyle | ViewStyle[]
}

export const Button = (props: ButtonProps) => {
    const { name, onPressButton, testId, rootStyle, ...rest} = props
    return(<TouchableOpacity testID={testId} onPress={onPressButton} style={[styles.root, rootStyle || []]}>
        <Text style={styles.txtName}>{name}</Text>
    </TouchableOpacity>)
}


const styles = StyleSheet.create({
    root: {
        backgroundColor:colors.primary,
        paddingHorizontal:20,
        paddingVertical: 10
    },
    txtName: {
        color: 'white',
        textAlign: "center"
    }
})