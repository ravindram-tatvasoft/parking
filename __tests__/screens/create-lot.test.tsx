// __tests__
import { cleanup, render, fireEvent, waitFor } from "@testing-library/react-native";
import * as React from "react";
import { Button, ButtonProps } from "../../components";
import MockedNavigator from "../../navigation/MockedNavigator";
import CreateLot from "../../screens/createlot/create-lot";
import { Provider } from 'react-redux';
import { store } from "../../redux/configureStore"
import * as Utils from '../../utils/utils'
import { setParkingSlotList, setTotalParking } from "../../redux/reducer/parking-slot-slice";


jest.mock('@react-navigation/native', () => ({
    ...jest.requireActual('@react-navigation/native'),
    useNavigation: () => ({ goBack: jest.fn() }),
    useRoute: () => ({
        params: {}
    }),
}));

jest.mock('@react-navigation/native', () => {
    const actualNav = jest.requireActual('@react-navigation/native');
    return {
        ...actualNav,
        useFocusEffect: () => jest.fn(),
        useNavigation: () => ({
            navigate: jest.fn(),
        }),
    };
});
let wrapper;

describe('Screen: CreateLot ', () => {

    beforeEach(() => {
        // wrapper = render(<MockedNavigator component={CreateLot}/>)
    })

    afterEach(() => {
        jest.clearAllMocks()
        cleanup()
    })

    test(`renders correctly`, () => {
        // const tree =  render(<MockedNavigator component={CreateLot} />)
        const navigation = { navigate: jest.fn() };
        const tree = render(
            <Provider store={store}>
                <CreateLot navigation={navigation} />
            </Provider>
        )
        expect(tree).toMatchSnapshot();
    });

    test("Enter lot should not be 0", async () => {
        const navigation = { navigate: jest.fn() };
        const { getByTestId, getByText, queryAllByText } = render(
            <Provider store={store}>
                <CreateLot navigation={navigation} />
            </Provider>
        )
        const totalPlot = "4";
        const createLotInput = getByTestId("CreateLot.inputPlot")
        await waitFor(() => {
            fireEvent.changeText(createLotInput, "4");
            expect(createLotInput.props.value).toBe(totalPlot);
        })
    })

    test("slotNumber should be > 0", async () => {
        const navigation = { navigate: jest.fn() };
        const { getByTestId, getByText, queryAllByText } = render(
            <Provider store={store}>
                <CreateLot navigation={navigation} />
            </Provider>
        )
        const totalPlot = "4";
        const createLotInput = getByTestId("CreateLot.inputPlot")
        await waitFor(() => {
            fireEvent.changeText(createLotInput, totalPlot);
            expect(createLotInput.props.value).toBe(totalPlot);
        })
    })

    test("should reset slot ", async () => {
        const navigation = { navigate: jest.fn() };
        const { getByTestId, getByText, queryAllByText } = render(
            <Provider store={store}>
                <CreateLot navigation={navigation} />
            </Provider>
        )
        const btnReset = getByTestId("idBtnResetSlot")
        const createLotInput = getByTestId("CreateLot.inputPlot")
        const createNoInput = getByTestId("CreateLot.inputCarNo")


        fireEvent.press(btnReset)
        await waitFor(() => {
            fireEvent.changeText(createLotInput, "");
            fireEvent.changeText(createNoInput, "");
        })
        expect(createLotInput.props.value).toBe("");
        expect(createNoInput.props.value).toBe("")
    })

    test("should open slot ", () => {
        const navigation = { navigate: jest.fn() };
        const { getByTestId, getByText, queryAllByText } = render(
            <Provider store={store}>
                <CreateLot navigation={navigation} />
            </Provider>
        )
        const btnOpenSlot = getByTestId("idBtnViewSlot")
        fireEvent.press(btnOpenSlot)
    })

    test("Car number must be there ", () => {
        const navigation = { navigate: jest.fn() };
        const { getByTestId, getByText, queryAllByText } = render(
            <Provider store={store}>
                <CreateLot navigation={navigation} />
            </Provider>
        );
        jest.spyOn(Utils, "validateCarNo").mockReturnValue(true)
        const mockCarNo = "GJ01XXXXXX"
        const inputCarNo = getByTestId("CreateLot.inputCarNo")
        fireEvent.changeText(inputCarNo, mockCarNo)
        expect(inputCarNo.props.value).toBe(mockCarNo)

    })


    test("Create slot: if 0 show error ", () => {
        const navigation = { navigate: jest.fn() };
        const { getByTestId, getByText, queryAllByText } = render(
            <Provider store={store}>
                <CreateLot navigation={navigation} />
            </Provider>
        );
        store.dispatch(setTotalParking(0))
        const btnCreateSlot = getByTestId("idBtnCreateSlot")
        fireEvent.press(btnCreateSlot)
    })

    test("Create slot: total slot > 0 ", async() => {
        const navigation = { navigate: jest.fn() };
        const { getByTestId, getByText, queryAllByText } = render(
            <Provider store={store}>
                <CreateLot navigation={navigation} />
            </Provider>
        );
        store.dispatch(setTotalParking(2))

        const createLotInput = getByTestId("CreateLot.inputPlot")
        await waitFor(() => {
            fireEvent.changeText(createLotInput, "4");
            expect(createLotInput.props.value).toBe("4");
        })

        const btnCreateSlot = getByTestId("idBtnCreateSlot")
        fireEvent.press(btnCreateSlot)
    })

    test("Add new car : valid number", async () => {
        const navigation = { navigate: jest.fn() };
        const { getByTestId, getByText, queryAllByText } = render(
            <Provider store={store}>
                <CreateLot navigation={navigation} />
            </Provider>
        );

        const inputCarNo = getByTestId("CreateLot.inputCarNo")
        const mockCarNo = "GJ 01 XX XXXX"
        await waitFor(() => {
            fireEvent.changeText(inputCarNo, mockCarNo)
        })
        expect(inputCarNo.props.value).toBe(mockCarNo)
        jest.spyOn(Utils, "validateCarNo").mockReturnValue(true)
        const btnAdd = getByTestId("idBtnEnterCar")
        fireEvent.press(btnAdd)
    })

    test("Add new car : Not valid number - show error", async () => {
        const navigation = { navigate: jest.fn() };
        const { getByTestId, getByText, queryAllByText } = render(
            <Provider store={store}>
                <CreateLot navigation={navigation} />
            </Provider>
        );

        const inputCarNo = getByTestId("CreateLot.inputCarNo")
        const mockCarNo = "GJ 01 XX XXX"
        await waitFor(() => {
            fireEvent.changeText(inputCarNo, mockCarNo)
        })
        expect(inputCarNo.props.value).toBe(mockCarNo)
        jest.spyOn(Utils, "validateCarNo").mockReturnValue(false)
        const btnAdd = getByTestId("idBtnEnterCar")
        fireEvent.press(btnAdd)
    })

    test("Add new car:: Slot not available show error", async () => {
        const navigation = { navigate: jest.fn() };
        const { getByTestId, getByText, queryAllByText } = render(
            <Provider store={store}>
                <CreateLot navigation={navigation} />
            </Provider>
        );

        const inputCarNo = getByTestId("CreateLot.inputCarNo")
        const mockCarNo = "GJ 01 XX XXXX"
        await waitFor(() => {
            fireEvent.changeText(inputCarNo, mockCarNo)
        })
        store.dispatch(setTotalParking(4))
        store.dispatch(setParkingSlotList([{
            isEmpty: false
        }]))
        expect(inputCarNo.props.value).toBe(mockCarNo)
        jest.spyOn(Utils, "validateCarNo").mockReturnValue(true)
        const btnAdd = getByTestId("idBtnEnterCar")
        fireEvent.press(btnAdd)
    })

    test("Add new car:: Slot must be available", async () => {
        const navigation = { navigate: jest.fn() };
        const { getByTestId, getByText, queryAllByText } = render(
            <Provider store={store}>
                <CreateLot navigation={navigation} />
            </Provider>
        );

        const inputCarNo = getByTestId("CreateLot.inputCarNo")
        const mockCarNo = "GJ 01 XX XXXX"
        await waitFor(() => {
            fireEvent.changeText(inputCarNo, mockCarNo)
        })
        store.dispatch(setTotalParking(4))
        store.dispatch(setParkingSlotList([{
            id: 1,
            isEmpty: true
        },
        {
            id: 2,
            isEmpty: false
        }]))
        expect(inputCarNo.props.value).toBe(mockCarNo)
        jest.spyOn(Utils, "validateCarNo").mockReturnValue(true)
        const btnAdd = getByTestId("idBtnEnterCar")
        fireEvent.press(btnAdd)


    })

})
