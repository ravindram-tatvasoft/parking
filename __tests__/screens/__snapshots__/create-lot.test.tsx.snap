// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen: CreateLot  renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "flex": 1,
        "width": "100%",
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    keyboardType="number-pad"
    onChangeText={[Function]}
    placeholder="Enter parking slot"
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "borderWidth": 1,
          "height": 40,
          "marginVertical": 12,
          "padding": 10,
        },
        Object {
          "marginHorizontal": 10,
        },
      ]
    }
    testID="CreateLot.inputPlot"
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#246EE9",
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
      testID="idBtnCreateSlot"
    >
      <Text
        style={
          Object {
            "color": "white",
            "textAlign": "center",
          }
        }
      >
        Create
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#246EE9",
          "marginHorizontal": 10,
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
      testID="idBtnResetSlot"
    >
      <Text
        style={
          Object {
            "color": "white",
            "textAlign": "center",
          }
        }
      >
        Reset Slot
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#246EE9",
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
      testID="idBtnViewSlot"
    >
      <Text
        style={
          Object {
            "color": "white",
            "textAlign": "center",
          }
        }
      >
        View Slot
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "backgroundColor": "black",
          "height": 1,
          "marginVertical": 25,
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "paddingHorizontal": 20,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="characters"
      keyboardType="default"
      onChangeText={[Function]}
      placeholder="Enter car no"
      rejectResponderTermination={true}
      style={
        Object {
          "borderWidth": 1,
          "height": 40,
          "marginVertical": 12,
          "padding": 10,
        }
      }
      testID="CreateLot.inputCarNo"
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#246EE9",
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
      testID="idBtnEnterCar"
    >
      <Text
        style={
          Object {
            "color": "white",
            "textAlign": "center",
          }
        }
      >
        Enter Car
      </Text>
    </View>
  </View>
</View>
`;
